Ensure that OpenStack cloud provider is initialized before making a call

Fedosin opened the pull request:
This commit improves the reading of OpenStack's cloud provider configuration from a secret. In some cases, if kubeConfigPath parameter specifies a configuration file, which is created by kubelet itself as a result of bootstrap (i.e. /var/lib/kubelet/kubeconfig), then the initialization of the cloud provider fails, because this file has not yet been created on the local file system. 
https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet-tls-bootstrapping/#bootstrap-initialization
To solve this problem we transfer the real initialization of the provider from the init() function, which is called at the stage of modules import, to the functions from the Provider's top-level interface. https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L42-L62
This allows to initialize the provider not at the stage of modules import, but later, when the requests to the functions occur, and the kubeconfig file is already created.

Another peculiarity of the provider is that in order to get the node names it needs to execute a request to OpenStack at the initial stage, when kubeconfig file is not created yet. As a solution, it is suggested not to make a request to OpenStack when reading the configuration from the secret and to give the nodes names that coincide with the hostnames. This approach is used by all other cloud providers, as well as the external provider for OpenStack. This solution will make it easier to migrate users from in-tree to external, as the node names will be the same in both cases.

/kind feature

```release-note
Ensure that OpenStack cloud provider initialized before making a call
```

derekwaynecarr commented:
As an operator that is wanting to ensure a smooth transition from in-tree to external cloud provider, we want to be able to have --cloud-provider set in the transition period, but ensure by policy that the node name matches the host-name.  WIth this patch, we are able to proceed through TLS bootstrapping, and the get access to the OpenStack secret to then operate as usual because the hostname and node name will match.  This seems like a good compromise to help operators like ourselves.

@dims wdyt?

derekwaynecarr commented:
basically, if TLS bootstrapping is enabled and OpenStack is configured with access via a secret, the OpenStack cloud provider cannot get the secret with the bootstrap config to properly configure, but if you know the node name will match the host name, it presents an unnecessary roadblock in bootstrap.

dims commented:
@hogepodge @chrigl @ramineni @adisky - does Derek's suggestion sound good to you?

mrunalp commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -294,19 +306,47 @@ func setConfigFromSecret(cfg *Config) error {
 	return fmt.Errorf("cannot find \"clouds.conf\" key in the secret")
 }
 
+func (os *OpenStack) ensureCloudProviderWasInitialized() error {
```
Should we cache the result of this function?

sjenning replied:
i think the result is "cached" by setting `os.provider` when the initialization succeeds (checked first thing in the next line).  correct me if i'm wrong.

mrunalp replied:
Yes :+1:

andrewsykim commented:
Curious if https://github.com/kubernetes/kubernetes/pull/75229 addresses this for OpenStack. In that PR the kubelet will try to discover it's internal IP and hostname if --cloud-provider=external and no addresses were previously set.

andrewsykim commented:
This is the logic the kubelet would follow when a node is initially created with no addresses set and --cloud-provider=external: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/nodestatus/setters.go#L150-L190

sjenning commented:
Took me a while to understand what was happening here but I get it now.  LGTM.

Fedosin commented:
/test pull-kubernetes-e2e-gce-100-performance

Fedosin commented:
/test pull-kubernetes-bazel-build

Fedosin commented:
@dims @hogepodge @chrigl @ramineni @adisky hi folks! I updated the pr description, now it's more clear what it does. and all the tests were passed.

andrewsykim commented on pkg/cloudprovider/providers/openstack/openstack_instances.go:
```diff
@@ -63,6 +71,11 @@ func (os *OpenStack) Instances() (cloudprovider.Instances, bool) {
 func (i *Instances) CurrentNodeName(ctx context.Context, hostname string) (types.NodeName, error) {
+	// if compute is not initialized just return hostname
+	if i.compute == nil {
+		return types.NodeName(hostname), nil
```
So is this local hostname override only required for kubelet TLS bootstrapping?

derekwaynecarr replied:
yes, which is consistent with how this will work w/ external cloud provider.

andrewsykim replied:
What do you think of https://github.com/kubernetes/kubernetes/pull/75229 where we try to discover the hostname and internal IPs if it wasn't already set by the external cloud provider. That would fix this problem for all cloud providers not just OpenStack.

Fedosin replied:
Yes, we override hostnames only for kubelet TLS bootstrapping if the config is taken from a secret.
There's an idea that for consistency, we should always return the hostname. What do you think?

andrewsykim replied:
I believe with https://github.com/kubernetes/kubernetes/pull/75229 we would return the hostname for all providers using the external providers.

Fedosin replied:
Yeah, #75229 makes sense to me and solves the issue for external cloud providers.
Here we are trying to do the same for in-tree OpenStack cloud-provider, which is the only one that gives out different names for hosts and nodes. 
So I think these patches complement each other.

andrewsykim replied:
I see, so I had the impression that this was mainly to help the transition to external providers since not setting hostname breaks TLS bootstrapping. So is TLS bootstrapping also broken for the in-tree case?

Fedosin commented:
@dims @hogepodge @chrigl @ramineni @adisky hey folks! do you have any comments about this patch?

andrewsykim commented:
Sorry for the delay, left a comment here https://github.com/kubernetes/kubernetes/pull/79221#discussion_r297213073!

adisky commented:
@Fedosin looks good to me

Fedosin committed:
Ensure that OpenStack cloud provider is initialized before making a call

When cloud provider reads data from a secret, there is no guarantee
that kubeconfig file exists when kubelet starts, and it may lead
to the fact that kubelet fails to operate.

This commit allows to ignore the absence of this file at the initial
stage of kubelet loading. Instead of making requests to the Nova servcer
it tries to read required data from the OpenStack metadata service.

Later on, when this file is in the file system, the provider is
initialized with the necessary configuration from the secret.

FengyunPan2 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

hogepodge commented:
As a heads up, one of these two patches will have to be rebased depending on order of landing. https://github.com/kubernetes/kubernetes/pull/80027
